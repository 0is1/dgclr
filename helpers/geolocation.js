// @flow
/**
 * Get current location
 * @return {Object} position object or error
 */

// eslint-disable-next-line import/prefer-default-export
export const getCurrentLocation = (getCurrentPositionSuccess, getCurrentPositionError) => {
  try {
    if (!navigator) return null;
    const options = {
      enableHighAccuracy: true,
      timeout: 10000,
      maximumAge: 0,
    };
    return navigator.geolocation.getCurrentPosition(getCurrentPositionSuccess, getCurrentPositionError, options);
  } catch (e) {
    console.error('getCurrentPosition error: ', e);
    throw e;
  }
};
