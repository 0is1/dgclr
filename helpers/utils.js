// @flow
import { uniq } from 'lodash';
import type { Layout, LocationInfo } from 'lib/types';
import matchAll from 'string.prototype.matchall';

/**
 * Convert links in text to <a href="">url</a> from first capture group
 * add "http://"-prefix if includeHttp is used
 * @param {String} text
 * @param {String} regex
 * @param {boolean} includeHttp
 * @return {String} links converted to <a>:s
 */
// eslint-disable-next-line max-len
export const removeLastCommaAndReturnLinkFromFirstCaptureGroup = (string: string, regex: any, includeHttp: boolean = false): string => string.replace(regex, (wholeMatch, group1) => {
  if (group1) {
    const includesLastDot = group1.endsWith('.');
    const returnString = includesLastDot ? group1.slice(0, -1) : group1;
    const hrefString = includeHttp ? `http://${returnString}` : returnString;
    return `<a href='${hrefString}'>${returnString}</a>${includesLastDot ? '.' : ''}`;
  }
  return '';
});

/**
 * Convert links in text to <a href="">url</a>
 * @param {String} text
 * @return {String} links converted to <a>:s
 */
export const LINK_REGEX = /((https?):\/\/[^\s/$.?#].[^\s)]*)/gim;
export const convertLinksToHtml = (string: string): string => removeLastCommaAndReturnLinkFromFirstCaptureGroup(string, LINK_REGEX);

/**
 * Convert coordinates array to {lat, lng} object
 * @param {Array} coordinates
 * @return {Object} {lat, lng}
 */
export const convertCoordinatesToObject = (coordinates: Array<number> = []): ?{ lat: number, lng: number } => {
  const [lng, lat] = coordinates;
  return parseFloat(lat) && parseFloat(lng) ? { lat, lng } : null;
};

/**
 * Get address details
 * @param {Object} locationInfo
 * @return {String} address details
 */
export const courseAddressDetails = (locationInfo: LocationInfo) => {
  const { address, city, zip } = locationInfo || {};
  if (address && zip && city) return `${address}, ${zip} ${city}`;
  if (zip && city) return `${zip} ${city}`;
  return '';
};

/**
 * Get unique layout ratings
 * @param {Array} layouts
 * @return {Array} ratings as strings
 */
// eslint-disable-next-line max-len
export const uniqueLayoutRatings = (layouts: Array<Layout> = []): Array<string> => uniq(layouts.filter((layout) => layout && layout.rating).map((layout) => layout.rating));

/**
 * Get unique key id
 * @return {String} unique id
 */

export const getRandomKey = () => Math.random()
  .toString(36)
  .substring(2, 15)
  + Math.random()
    .toString(36)
    .substring(2, 15);

/**
 * Get page title
 * @param {String} text
 * @return {String} title
 */

export const getTitle = (text: ?string = null) => (text ? `DGCLR - ${text}` : 'DGCLR - Disc Golf Course Lists and Results');

/**
 * Are we in browser
 * @return {boolean}
 */
export const isBrowser = typeof window !== 'undefined';

/**
 * Is array with length
 * @param {Array} array
 * @return {boolean}
 */
export const isArrayWithLength = (array: ?Array<any>): boolean => Array.isArray(array) && array.length > 0;

/**
 * Convert meters to kilometers
 * @param {Long} meters
 * @return {int}
 */
export const convertMetersToKilometers = (meters: number): number => parseInt(meters / 1000, 10) || 0;

/**
 * Convert kilometers to meters
 * @param {Long} kilometers
 * @return {int}
 */
export const convertKilometersToMeters = (kilometers: number): number => parseInt(kilometers * 1000, 10) || 0;

/**
 * Get course map url from layouts based on index
 * @param {Array} layouts
 * @return {string}
 */
export const getCourseMapUrlForLayout = (layouts: Array<Layout> = [], activeIndex: number): string => (layouts && layouts[activeIndex] && layouts[activeIndex].mapUrl) || '';

export const WWW_REGEX = /(www\.[a-öA-Ö]*\.([a-öA-Ö]*)[/a-öA-Ö?&0-9=.]*)/gi;
const HTTP_REGEX = /(https?):\/\/([^\s/$.?#].[^\s<>'")]*)/gi;
/**
 * Convert text with only "www" to links in text to <a href="">url</a>
 * @param {String} text
 * @return {String} links converted to <a>:s
 */
export const convertWWWToHttpAndAddLinks = (string: string): string => {
  if (!isArrayWithLength(string.match(WWW_REGEX))) {
    return string;
  }
  const httpMatches = Array.from(matchAll(string, HTTP_REGEX));
  // console.log('httpMatches: ', httpMatches);
  const wwwMatches = Array.from(new Set(string.match(WWW_REGEX))).map((item) => (item.endsWith('.') ? item.slice(0, -1) : item));
  // console.log('wwwMatches: ', wwwMatches);
  if (isArrayWithLength(httpMatches) && isArrayWithLength(wwwMatches)) {
    let stringWithWWWLinks = string;
    const ignoredWWWTexts = httpMatches.map((matchArray) => {
      // www-match is second group => third item in the array
      const [, , wwwMatch] = matchArray;
      // Remove last dot if there is one
      return wwwMatch.endsWith('.') ? wwwMatch.slice(0, -1) : wwwMatch;
    });
    // console.log('ignoredWWWTexts: ', ignoredWWWTexts);
    // If this www.something.com is already wrapped with <a>, filter it out from wwwTextsWithoutLink
    const wwwTextsWithoutLink = wwwMatches.filter((www) => !ignoredWWWTexts.includes(www));
    // console.log('wwwTextsWithoutLink: ', wwwTextsWithoutLink);
    wwwTextsWithoutLink.forEach((wwwString) => {
      const regex = new RegExp(`(${wwwString})`, 'g');
      stringWithWWWLinks = stringWithWWWLinks.replace(regex, "<a href='http://$1'>$1</a>");
    });
    return stringWithWWWLinks;
  }
  return removeLastCommaAndReturnLinkFromFirstCaptureGroup(string, WWW_REGEX, true);
};
