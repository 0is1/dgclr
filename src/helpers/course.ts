import { Course } from '../types';

type SearchCourseBySlug = {
  courseBySlug: Course[];
};
export const getCourseDataFromSearchCourseBySlug = (
  data?: SearchCourseBySlug
) => {
  if (!data || !data.courseBySlug) {
    return undefined;
  }
  return data.courseBySlug[0];
};

export const getCourseNameFromCourseData = (data?: Course) => {
  if (!data) {
    return null;
  }
  return data.name;
};

export const getCourseAddressFromCourseData = (data?: Course) => {
  if (!data?.locationInfo) {
    return '';
  }
  const { address, city, zip } = data.locationInfo;
  if (!address || !city || !zip) {
    return '';
  }
  return `${address}, ${zip} ${city}`;
};

export const getCourseCoordinatesFromCourseData = (data?: Course) => {
  if (!data?.locationInfo?.location?.coordinates) {
    return [];
  }
  return data.locationInfo.location.coordinates;
};

export const getLocationCoordinatesFromCourseData = (data?: Course) => {
  if (!data?.locationInfo?.location?.coordinates) {
    return undefined;
  }
  return data.locationInfo.location.coordinates;
};

export const getCourseLayoutImage = (
  course: Course | undefined,
  layoutIndex: string
) => {
  if (!course) {
    return '';
  }
  const layout = course.layouts[parseInt(layoutIndex, 10)];
  if (!layout) {
    return '';
  }
  return layout.mapUrl;
};

export const getTabListFromCourseLayouts = (course: Course | undefined) => {
  if (!course || !course.layouts) {
    return [];
  }
  return course.layouts.map((layout, index) => {
    return {
      key: index.toString(),
      tab: layout.name,
    };
  });
};
export const getRatingListFromCourseLayouts = (course: Course | undefined) => {
  if (!course || !course.layouts) {
    return [];
  }
  return course.layouts
    .map((layout, index) => {
      return {
        key: index.toString(),
        rating: layout.rating,
      };
    })
    .filter((layout) => layout.rating);
};

export const getCourseLayoutByIndex = (
  course: Course | undefined,
  layoutIndex: string
) => {
  if (!course || !course.layouts) {
    return null;
  }
  return course.layouts[parseInt(layoutIndex, 10)];
};

export const getCourseInfo = (course: Course | undefined) => {
  if (!course) {
    return null;
  }
  return {
    ...course.courseInfo,
    ...course.locationInfo,
  };
};

export const omitHiddenLayoutsFromCourses = (courses: Course[]) => {
  const coursesWithoutHiddenLayouts = courses.map((course) => {
    const { layouts } = course;
    if (!layouts?.length) return course;
    const layoutsWithoutHidden = layouts.filter((layout) => !layout.hidden);
    return { ...course, layouts: layoutsWithoutHidden };
  });
  return coursesWithoutHiddenLayouts;
};

export const getCourseInitialNameFromSlug = (slug: string) => {
  // slug is in the format of "course-name-city" like "golf-club-berlin"
  const initialName = slug.split('-').slice(0, -1).join(' ');
  // convert first word first letter to uppercase
  return initialName.charAt(0).toUpperCase() + initialName.slice(1);
};
