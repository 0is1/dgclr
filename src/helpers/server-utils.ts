export const getRequestHeaders = () => ({
  authorization: `Bearer ${process.env.BEARER_TOKEN}`,
});
