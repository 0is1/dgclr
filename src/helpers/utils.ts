/**
 * Convert links in text to <a href="">url</a>

 */
export const LINK_REGEX = /((https?):\/\/[^\s/$.?#].[^\s)]*)/gim;
export const convertLinksToHtml = (string?: string): string =>
  string
    ? removeLastCommaAndReturnLinkFromFirstCaptureGroup(string, LINK_REGEX)
    : "";

/**
 * Is array with length
 */
export const isArrayWithLength = (array?: any): boolean =>
  Array.isArray(array) && array.length > 0;

/**
 * Convert links in text to <a href="">url</a> from first capture group
 * add "http://"-prefix if includeHttp is used
 */
// eslint-disable-next-line max-len
export const removeLastCommaAndReturnLinkFromFirstCaptureGroup = (
  string: string,
  regex: any,
  includeHttp: boolean = false
): string =>
  string.replace(regex, (wholeMatch, group1) => {
    if (group1) {
      const includesLastDot = group1.endsWith(".");
      const returnString = includesLastDot ? group1.slice(0, -1) : group1;
      const hrefString = includeHttp ? `http://${returnString}` : returnString;
      return `<a href='${hrefString}'>${returnString}</a>${
        includesLastDot ? "." : ""
      }`;
    }
    return "";
  });

export const WWW_REGEX = /(www\.[a-öA-Ö]*\.([a-öA-Ö]*)[/a-öA-Ö?&0-9=.]*)/gi;
const HTTP_REGEX = /(https?):\/\/([^\s/$.?#].[^\s<>'")]*)/gi;
/**
 * Convert text with only "www" to links in text to <a href="">url</a>
 */
export const convertWWWToHttpAndAddLinks = (string: string): string => {
  if (!isArrayWithLength(string.match(WWW_REGEX))) {
    return string;
  }
  const httpMatches = Array.from(string.matchAll(HTTP_REGEX));
  // console.log('httpMatches: ', httpMatches);
  const wwwMatches = Array.from(new Set(string.match(WWW_REGEX))).map((item) =>
    item.endsWith(".") ? item.slice(0, -1) : item
  );
  // console.log('wwwMatches: ', wwwMatches);
  if (isArrayWithLength(httpMatches) && isArrayWithLength(wwwMatches)) {
    let stringWithWWWLinks = string;
    const ignoredWWWTexts = httpMatches.map((matchArray) => {
      // www-match is second group => third item in the array
      const [, , wwwMatch] = matchArray;
      // Remove last dot if there is one
      return wwwMatch.endsWith(".") ? wwwMatch.slice(0, -1) : wwwMatch;
    });
    // console.log('ignoredWWWTexts: ', ignoredWWWTexts);
    // If this www.something.com is already wrapped with <a>, filter it out from wwwTextsWithoutLink
    const wwwTextsWithoutLink = wwwMatches.filter(
      (www) => !ignoredWWWTexts.includes(www)
    );
    // console.log('wwwTextsWithoutLink: ', wwwTextsWithoutLink);
    wwwTextsWithoutLink.forEach((wwwString) => {
      const regex = new RegExp(`(${wwwString})`, "g");
      stringWithWWWLinks = stringWithWWWLinks.replace(
        regex,
        "<a href='http://$1'>$1</a>"
      );
    });
    return stringWithWWWLinks;
  }
  return removeLastCommaAndReturnLinkFromFirstCaptureGroup(
    string,
    WWW_REGEX,
    true
  );
};

export const getBrowserPrefersDarkMode = (): boolean => {
  if (typeof window !== "undefined") {
    return window.matchMedia("(prefers-color-scheme: dark)").matches;
  }
  return false;
};
