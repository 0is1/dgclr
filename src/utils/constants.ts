export const DEFAULT_LIMIT = 10;

export const RATING_OPTIONS = [
  { value: 'AAA1', label: 'AAA1' },
  { value: 'AA1', label: 'AA1' },
  { value: 'A1', label: 'A1' },
  { value: 'AA2', label: 'AA2' },
  { value: 'AA3', label: 'AA3' },
  { value: 'A2', label: 'A2' },
  { value: 'A3', label: 'A3' },
  { value: 'BB1', label: 'BB1' },
  { value: 'BB2', label: 'BB2' },
  { value: 'B1', label: 'B1' },
  { value: 'B2', label: 'B2' },
  { value: 'B3', label: 'B3' },
  { value: 'C1', label: 'C1' },
  { value: 'C2', label: 'C2' },
  { value: 'C3', label: 'C3' },
  { value: 'D1', label: 'D1' },
  { value: 'D2', label: 'D2' },
  { value: 'D3', label: 'D3' },
];

export const SURFACE_TYPE_OPTIONS = [
  {
    value: 'Mäkinen, paljon korkeuseroja',
    label: 'Mäkinen, paljon korkeuseroja',
  },
  {
    value: 'Kumpuileva, jonkin verran korkeuseroja',
    label: 'Kumpuileva, jonkin verran korkeuseroja',
  },
  {
    value: 'Tasainen, vähän korkeuseroja',
    label: 'Tasainen, vähän korkeuseroja',
  },
  { value: 'Tasainen, ei korkeuseroja', label: 'Tasainen, ei korkeuseroja' },
];

export const COURSE_TYPE_OPTIONS = [
  {
    value: 'Metsärata',
    label: 'Metsärata',
  },
  {
    value: 'Puistorata',
    label: 'Puistorata',
  },
  {
    value: 'Peltorata',
    label: 'Peltorata',
  },
  {
    value: 'Hiekkakuopparata',
    label: 'Hiekkakuopparata',
  },
  {
    value: 'Hiihtokeskus',
    label: 'Hiihtokeskus',
  },
];
