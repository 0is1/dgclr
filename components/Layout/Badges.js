// @flow
import React from 'react';
import { Badge } from 'rebass';
import { getRandomKey } from 'helpers/utils';
import type { Node } from 'react';

type Props = {
  ratings: Array<string>,
  tiny?: boolean,
};

const LayoutRatingBadges = ({ ratings, tiny }: Props) => {
  const badges: Array<?Node> = ratings.map((rating) => {
    if (rating) {
      return tiny ? (
        <Badge mx="8px" py={0} key={getRandomKey()}>
          {rating}
        </Badge>
      ) : (
        <Badge key={getRandomKey()}>{rating}</Badge>
      );
    }
    return null;
  });
  return badges;
};

LayoutRatingBadges.defaultProps = {
  tiny: false,
};

export default LayoutRatingBadges;
