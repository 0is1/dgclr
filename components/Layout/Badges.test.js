/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import { Badge } from 'rebass';
import { mount, shallow } from 'enzyme';
import mockCoursesData from 'components/SearchContainer/mock/courses.mock';
import { uniqueLayoutRatings } from 'helpers/utils';
import Badges from './Badges';

const defaultProps = {
  tiny: false,
  ratings: uniqueLayoutRatings(mockCoursesData[1].layouts),
};

describe('Layout component', () => {
  it('Match snapshot', () => {
    const component = renderer.create(<Badges {...defaultProps} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('Renders correctly without ratings', () => {
    const enzymeWrapper = shallow(<Badges ratings={[]} />);
    expect(enzymeWrapper).toBeDefined();
    expect(enzymeWrapper.find(Badge).length).toEqual(0);
  });
  it('Renders correctly', () => {
    const enzymeWrapper = mount(<Badges {...defaultProps} />);
    expect(enzymeWrapper).toBeDefined();
    expect(enzymeWrapper.find(Badge).length).toEqual(1);
    expect(enzymeWrapper.find(Badge).text()).toEqual('AA1');
  });
});
