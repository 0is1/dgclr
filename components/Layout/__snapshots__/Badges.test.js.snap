// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout component Match snapshot 1`] = `
<div
  className="sc-fAjcbJ chDNXh"
>
  AA1
</div>
`;
